### PR TITLE
[wip] Add an option for Caffe2 to link with local protobuf.

### PR DESCRIPTION
**Pull request by Yangqing**
Continuation of the protobuf hiding work following #2302 . Starting a PR to test build.

### CODE REVIEW

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -76,12 +76,38 @@ install(FILES ${PROJECT_BINARY_DIR}/caffe2/core/macros.h
...
 add_library(caffe2_protos $<TARGET_OBJECTS:Caffe_PROTO> $<TARGET_OBJECTS:Caffe2_PROTO>)
 add_dependencies(caffe2_protos Caffe_PROTO Caffe2_PROTO)
 # TODO: enable using lite protobuf.
-target_link_libraries(caffe2_protos PUBLIC protobuf::libprotobuf)
+# If we are going to link protobuf locally inside caffe2 libraries, what we will do is
+# to create a helper static library that always contains libprotobuf source files, and
+# link the caffe2 related dependent libraries to it.
 target_include_directories(caffe2_protos INTERFACE $<INSTALL_INTERFACE:include>)
+# Reason for this public dependency is as follows:
+# (1) Strictly speaking, we should not expose any Protobuf related functions. We should
```
Comment (orionr) [inline caffe2/CMakeLists.txt:87]: Completely agree on this. Great changes.

File: caffe2/CMakeLists.txt [hunk 2]
```diff
@@ -76,12 +76,38 @@ install(FILES ${PROJECT_BINARY_DIR}/caffe2/core/macros.h
...
+#     only use function interfaces wrapped with our own public API, and link protobuf
+#     locally.
+# (2) However, currently across the Caffe2 codebase, we have extensive use of protobuf
```
Comment (orionr) [inline caffe2/CMakeLists.txt:90]: Is the python extension taken care of with these changes? Just checking.

**Review by orionr — COMMENTED** [review]
This a *great* change. cc @pjh5

**Review by pjh5 — APPROVED** [review]

File: caffe2/CMakeLists.txt (hunk 2 above)
Comment (Yangqing) [reply caffe2/CMakeLists.txt:90]: Currently no - it still depends on the transitive dependency of libprotobuf static. But no explicit dependency on any libprotobuf.so library is needed now.

**Pull request merged**